<a href="https://colab.research.google.com/github/Abrazacs/linear_algebra/blob/main/%D0%AD%D0%BA%D0%B7%D0%B0%D0%BC%D0%B5%D0%BD%D0%B0%D1%86%D0%B8%D0%BE%D0%BD%D0%BD%D0%BE%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_%D0%BB%D0%B8%D0%BD%D0%B5%D0%B9%D0%BD%D0%B0%D1%8F_%D0%B0%D0%BB%D0%B3%D0%B5%D0%B1%D1%80%D0%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Часть 1**

**Задание 1**

Исследуйте систему векторов на линейную зависимость или независимость без использования Python (вручную):

⃗a1(1,8,-1)  

⃗a2(-2,3,3)

⃗a3(4,-11,9)

Подробно распишите решение, а в итоге сделайте вывод про то, являются ли векторы зависимыми или независимыми.


---

**Решение**


Представим векторы в виде матрицы


\begin{pmatrix}
1 & -2 & 4 \\
8 & 3 & -11 \\
-1 & 3 & 9
\end{pmatrix}

Для проверки линейной независимости вычислим ранг этой матрицы. Если ранг матрицы равен числу векторов (т.е. 3), то векторы линейно независимы. Если ранг меньше 3, то векторы линейно зависимы.

Приведем ее к ступенчатому виду.

Вычтем первую строку из второй (предварительно умножив на 8), а также
сложим первую и третью сторки

\begin{pmatrix}
1 & -2 & 4 \\
0 & 19 & -43 \\
0 & 1 & 13
\end{pmatrix}

Вторую сторку делим на 19 и вычетаем из третьей

\begin{pmatrix}
1 & -2 & 4 \\
0 & 19 & -43 \\
0 & 0 & 15
\end{pmatrix}

*15 - это округленное значение*

Ранг ступенчатой матрицы равен количеству её ненулевых строк

Т.к. все строки не нулевые, то ранг = 3.

Следовательно векторы линейно независимы.

**Задание 2**

При каких значениях параметра a система векторов является линейно зависимой? Найдите решение без использования Python (вручную).

⃗a1(1,0,2,-3)  

⃗a2(2,2,-4,a)

⃗a3(3,1,2,-5)


---

Решение

Представим векторы в виде матрицы

\begin{pmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2 \\
-3 & a & -5
\end{pmatrix}

Найдем опдределитель подматрицы

\begin{pmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2
\end{pmatrix}


det = 1 * (2 * 2 - 1 * -4) - 2 * (0 * 2 - 1 * 2) + 3 * (0 * - 4 - 2 * 2) =
8 + 4 - 12 = 0

**Ответ:** Т.к. определитель равен 0 , то это значит, что вектора линейнозависимые при любом a

**Задание 3**

Выразите определитель матрицы. Найдите решение без использования Python (вручную).

\begin{pmatrix}
1 & -1 & a & -1 \\
-1 & -2 & b & -1 \\
-2 & 0 & c & 1 \\
0 & 1 & d & 0
\end{pmatrix}


---


**Решение:**

det(A) = -0 * det(A41) + det(A42) -d * det(A43) + 0 * det(A44)

det(A42) = 3a + 2c - b

det(A43) = -1

det(A) = 3a + 2c - b + d

**Ответ:**

3a + 2c - b + d

**Задание 4**

Найдите значения параметров a, b и c, при которых матрицы A и B являются обратными. Найдите решение без использования Python (вручную).

Матрица A

\begin{pmatrix}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2c
\end{pmatrix}

Матрица B

\begin{pmatrix}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{pmatrix}

Две матрицы являются обратными друг другу, если их произведение равно единичной матрице. То есть, A * B = E.

A * B = \begin{pmatrix}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2c
\end{pmatrix} * \begin{pmatrix}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{pmatrix}

= \begin{pmatrix}
-4a+5 & 3a-4+1 & 5a-6+1 \\
60-60 & -45+2b+40-12 & -75+3b+60-12 \\
-40-10c & 30-38-2c & 50-57-2c
\end{pmatrix}

Эта матрица должна быть равна единичной матрице:

\begin{pmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1
\end{pmatrix}

Получаем систему уравнений:

-4a+5 = 1

3a-3 = 0

5a-5 = 0

2b-17 = 1

3b-27 = 0

-40-10c = 0

-8-2c = 0

-7-2c = 1

**Решение:**

a = 1

b = 9

c = -4

**Задание 5**

Найдите собственные векторы и собственные значения матрицы. Найдите решение без использования Python (вручную).

\begin{pmatrix}
4 & 0 & -1 \\
2 & 2 & 1 \\
-1 & 0 & 4
\end{pmatrix}


---

det(A - λE) = det\begin{pmatrix}
4-λ & 0 & -1 \\
2 & 2-λ & 1 \\
-1 & 0 & 4-λ
\end{pmatrix} = 0

-λ^3+10λ^2-31λ+30 = -(λ-2)(λ^2-8λ+15) = -(λ-2)(λ-3)(λ-5)

λ1 = 2

λ2 = 3

λ3 = 5

Подставляем λ1 = 2

\begin{pmatrix}
2 & 0 & -1 \\
2 & 0 & 1 \\
-1 & 0 & 2
\end{pmatrix}

Получаем систему

2x - z = 0

2x + z = 0

-x + 2z = 0

z = 0

x = 0

y может быть любым, т.к. коэффициент при y = 0

v1 = (0, 5, 0)

Подставляем λ2 = 3

\begin{pmatrix}
1 & 0 & -1 \\
2 & -1 & 1 \\
-1 & 0 & 1
\end{pmatrix}

Получаем систему уравнений

x - z = 0
2x - y + z = 0
-x + z = 0

x = z

y = 3z

Подставим любое значение. К примеру 2

v2 = (2,6,2)

Подставляем λ3 = 5

\begin{pmatrix}
-1 & 0 & -1 \\
2 & -3 & 1 \\
-1 & 0 & -1
\end{pmatrix}

Получаем систему

-x - z = 0

2x -3y + z = 0

-x = z

2x - 3y -x = 0

x - 3y = 0

y = x/3

z = -x

Подставим 1 в качестве x

v3 = (1, 1/3, -1)

**Ответ:**

v1 = (0, 5, 0)

v2 = (2, 6, 2)

v3 = (1, 1/3, -1)

Задание 6

Найдите сумму сингулярных чисел для матрицы:

\begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix}

Найдите решение без использования Python (вручную).

---
A = \begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix}


A^T * A = \begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix} * \begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix} =
\begin{pmatrix}
5 & 1 & -3 \\
1 & 5 & -3 \\
-3 & -3 & 3
\end{pmatrix}

Находим собственные значения матрицы A^T * A

det\begin{pmatrix}
5-λ & 1 & -3 \\
1 & 5-λ & -3 \\
-3 & -3 & 3-λ
\end{pmatrix} = 0

Находим характирестическое уравнение

(5-λ) * det\begin{pmatrix}
5-λ & -3 \\
-3 & 3-λ
\end{pmatrix} - 1 * det\begin{pmatrix}
1 & -3 \\
-3 & 3-λ
\end{pmatrix} + (-3) * det\begin{pmatrix}
1 & 5-λ \\
-3 & -3
\end{pmatrix}

λ^3 - 13λ^2 + 36λ = 0

λ(λ^2 - 13λ + 36) = 0

λ1 = 0

λ^2 - 13λ + 36 = 0

λ2,3 = (13 ± sqrt((-13)^2 - 4 * 1 * 36)) / 2 * 1

λ2 = 4

λ3 = 9

Сингулярные числа матрицы A - это квадратные корни из собственных значений матрицы A^T * A:

σ1 = sqrtλ1 = 0

σ2 = sqrtλ2 = 2

σ3 = sqrtλ3 = 3

Ответ: 5


# **Часть 2: Задания, решаемые с помощью Python**

Задание 7

Загрузите данные. Изучите, есть ли в них пропущенные значения?Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [27]:
import pandas as pd
import numpy as np
import sklearn.metrics as metrics


df = pd.read_csv('asset-v1_SkillFactory+MIFIML-1sem+2024+type@asset+block@bikes_rent.xls', sep=',', encoding='utf-8')
df.head()

season  yr  mnth  holiday  weekday  workingday  weathersit       temp  \
0       1   0     1        0        6           0           2  14.110847   
1       1   0     1        0        0           0           2  14.902598   
2       1   0     1        0        1           1           1   8.050924   
3       1   0     1        0        2           1           1   8.200000   
4       1   0     1        0        3           1           1   9.305237   

      atemp      hum  windspeed(mph)  windspeed(ms)   cnt  
0  18.18125  80.5833       10.749882       4.805490   985  
1  17.68695  69.6087       16.652113       7.443949   801  
2   9.47025  43.7273       16.636703       7.437060  1349  
3  10.60610  59.0435       10.739832       4.800998  1562  
4  11.46350  43.6957       12.522300       5.597810  1600

In [28]:
df.isnull().sum()

season            0
yr                0
mnth              0
holiday           0
weekday           0
workingday        0
weathersit        0
temp              0
atemp             0
hum               0
windspeed(mph)    0
windspeed(ms)     0
cnt               0
dtype: int64

Пропущенные значения отсутствуют

**Задание 8**

Исследуем данные на наличие линейной зависимости между столбцами. Найдите ранг матрицы данных. Сравните его с количеством столбцов, сделайте вывод. Если линейная зависимость есть - напишите, между какими столбцами и удалите один из них.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [29]:
df.corr()

season        yr      mnth   holiday   weekday  workingday  \
season          1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr             -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth            0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday        -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday        -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday      0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit      0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp            0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp           0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum             0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(mph) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
windspeed(ms)  -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt             0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

                weathersit      temp     atemp       hum  windspeed(mph)  \
season            0.019211  0.334315  0.342876  0.205445       -0.229046   
yr               -0.048727  0.047604  0.046106 -0.110651       -0.011817   
mnth              0.043528  0.220205  0.227459  0.222204       -0.207502   
holiday          -0.034627 -0.028556 -0.032507 -0.015937        0.006292   
weekday           0.031087 -0.000170 -0.007537 -0.052232        0.014282   
workingday        0.061200  0.052660  0.052182  0.024327       -0.018796   
weathersit        1.000000 -0.120602 -0.121583  0.591045        0.039511   
temp             -0.120602  1.000000  0.991702  0.126963       -0.157944   
atemp            -0.121583  0.991702  1.000000  0.139988       -0.183643   
hum               0.591045  0.126963  0.139988  1.000000       -0.248489   
windspeed(mph)    0.039511 -0.157944 -0.183643 -0.248489        1.000000   
windspeed(ms)     0.039511 -0.157944 -0.183643 -0.248489        1.000000   
cnt              -0.297391  0.627494  0.631066 -0.100659       -0.234545   

                windspeed(ms)       cnt  
season              -0.229046  0.406100  
yr                  -0.011817  0.566710  
mnth                -0.207502  0.279977  
holiday              0.006292 -0.068348  
weekday              0.014282  0.067443  
workingday          -0.018796  0.061156  
weathersit           0.039511 -0.297391  
temp                -0.157944  0.627494  
atemp               -0.183643  0.631066  
hum                 -0.248489 -0.100659  
windspeed(mph)       1.000000 -0.234545  
windspeed(ms)        1.000000 -0.234545  
cnt                 -0.234545  1.000000

In [30]:
rank = np.linalg.matrix_rank(df.select_dtypes(include=[np.number]))
print(f"Ранг матрицы: {rank}")

Ранг матрицы: 12


Кол-во столбцов в матрице = 13, а ранг матрицы = 12 следовательно есть линейнозависимые столбцы. Также это отражено в таблице корреляции - столбцы windspeed(mph)	windspeed(ms).

Удалим столбец windspeed(mph)

In [31]:
df.drop(['windspeed(mph)'], axis=1, inplace=True)

**Задание 9**

Выведите корреляционную матрицу оставшихся переменных.

* Какая переменная сильнее всего коррелирует с целевой?

* Есть ли мультиколлинеарность? Между какими переменными? Если есть корреляция выше 0.9 - оставьте только 1 признак. Объясните выбор

In [32]:
df.corr()

season        yr      mnth   holiday   weekday  workingday  \
season         1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr            -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth           0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday       -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday       -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday     0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit     0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp           0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp          0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum            0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(ms) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt            0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

               weathersit      temp     atemp       hum  windspeed(ms)  \
season           0.019211  0.334315  0.342876  0.205445      -0.229046   
yr              -0.048727  0.047604  0.046106 -0.110651      -0.011817   
mnth             0.043528  0.220205  0.227459  0.222204      -0.207502   
holiday         -0.034627 -0.028556 -0.032507 -0.015937       0.006292   
weekday          0.031087 -0.000170 -0.007537 -0.052232       0.014282   
workingday       0.061200  0.052660  0.052182  0.024327      -0.018796   
weathersit       1.000000 -0.120602 -0.121583  0.591045       0.039511   
temp            -0.120602  1.000000  0.991702  0.126963      -0.157944   
atemp           -0.121583  0.991702  1.000000  0.139988      -0.183643   
hum              0.591045  0.126963  0.139988  1.000000      -0.248489   
windspeed(ms)    0.039511 -0.157944 -0.183643 -0.248489       1.000000   
cnt             -0.297391  0.627494  0.631066 -0.100659      -0.234545   

                    cnt  
season         0.406100  
yr             0.566710  
mnth           0.279977  
holiday       -0.068348  
weekday        0.067443  
workingday     0.061156  
weathersit    -0.297391  
temp           0.627494  
atemp          0.631066  
hum           -0.100659  
windspeed(ms) -0.234545  
cnt            1.000000

Самая сильная корреляция целевой переменной cnt с переменной atemp = 0.631066

Между переменными temp и atemp имеется высокая корреляция 0.991702. Это является признаком мультиколлинеарности.

Удалис atemp, т.к. информация здесь является менее точной, чем в столбце temp, несмотря на то, что корреляция atemp с целевым признаком выше, чем у temp

In [33]:
df.drop(['atemp'], axis=1, inplace=True)

**Задание 10**

Разделите данные на предикторы и целевую переменную. Обучите модель с помощью классической линейной регрессии. Линейную регрессию необходимо реализовать без готовых библиотек.

Оцените качество модели с помощью хотя бы двух метрик качества.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [34]:
X = df.drop(['cnt'], axis=1)
y = df['cnt']

X = np.c_[np.ones(X.shape[0]), X]

X = np.array(X)
y = np.array(y)

coefficients = np.linalg.inv(X.T @ X) @ X.T @ y

y_pred = X @ coefficients

print(f'MAE: {metrics.mean_absolute_error(y, y_pred)}')
print(f'MSE: {metrics.mean_squared_error(y, y_pred)}')
print(f'R2: {metrics.r2_score(y, y_pred)}')



MAE: 648.7802547417004
MSE: 753878.6752929662
R2: 0.7988398641466421
